siod/siod.1
.TH SIOD 1C LOCAL
.SH NAME
siod \- small scheme interpreter (Scheme In One Defun).
.SH SYNOPSIS
.B siod
[-sX] [-hX] [-iX] [-gX] [-oX] [-nX] [-eX] [-vX] [-lX] [-mX] [-vX] [filenames]
.SH DESCRIPTION
.I Siod
is a very small scheme interpreter which can be used for
interactive calculations,
scripts, or included as a command interpreter or extension/macro
language in other applications. See the documentation for interfacing
requirements and how to add user-defined data types. The dash-style command
line options may also be delimited by commas.

.RE
.SH ARGUMENTS
The arguments are files which are loaded from the current or SIOD_LIB
directory, by usual default /usr/local/lib/siod. Or the symbol repl
to cause entry into the read-eval-print-loop, or parenthetical
expressions to be evaluated.

.RE
.SS COMMAND LINE OPTIONS
.TP 8
.BI \-h "XXXXX"
The
.I XXXXX
should be an integer, specifying the number of cons cells to
allocate in the heap. The default is 100000:10. Or an integer followed by
a colon, the number after which gives the maximum number of heaps to
malloc with mark and sweep gc mode.
.TP
.BI \-i "XXXXX"
The
.I XXXXX
should be the name of an init file to load before going into
the read/eval/print loop.
.TP
.BI \-g "X"
The
.I X
is 1 for a stop and copy garbage collector, 0 for a mark
and sweep one (the default).
.TP
.BI \-o "XXXXX"
The
.I XXXXX
should be an integer, specifying the size of the obarray (symbol hash table)
to use. Defaults to 1000. Each array element is a list of symbols.
.TP
.BI \-n "XXXXX"
The
.I XXXXX
should be an integer, specifying the number of pre-cons numbers
to create. The default is 100.
.TP
.BI \-s "XXXXX"
The
.I XXXXX
should be an integer, specifying the number of bytes of recursion
on the machine (C-call frame) stack to allow. This may be changed
while the programming is running, and is mainly a convenience for
detecting defects in programs. Default 200000.
.TP
.BI \-e "XXXXX"
The
.I XXXXX
is an expression to evaluate (after loading the init file, if any).
After evaluating the expression the program will exit.
.TP
.BI \-l "XXXXX"
The
.I XXXXX
is a library directory to use for the require procedure. defaults
to the environment variable SIOD_LIB or /usr/local/lib/siod.
.TP
.BI \-v "XXXXX"
The
.I XXXXX
is an integer verboseness from 0 to 4. When used in scripts the
most generally useful level is 0. Defaults to 4.
.TP 8
.BI \-m "XXXXX"
The
.I XXXXX
can be 0 (default) which has no effect, or 1, which causes the scheme
procedure main, with no arguments to be invoked after all files are loaded,
or 2, which does the same thing but prevents further arguments from
being interpreted, or 3, which will wrap special error handling
around the call to main which is convenient for writing cgi scripts
producing html. The environment may be obtained from the getenv
procedure or the variable *env*. Arguments are found in *args*.
.TP
.BI \-v "XXXXX"
The
.I XXXXX
Is passed along to the siod argument processing, however if the value
is greater than 0 a Content-type: text/plain header is generated
immediately (except if the number is represented with a leading zero),
which can be convenient for debugging WWW cgi scripts. For example:

.EX
#!/usr/local/bin/siod -v0,-m2
(define (main)
  (writes nil "Hello World\n"))
.EE

Which if invoked normally will result in

.EX
%hello.scm
Hello World
.EE

And with the verbose flag over-ride on the command line:

.EX

%hello.scm -v5
Content-type: text/plain

Welcome to SIOD, Scheme In One Defun, Version 3.1x 12-JUN-95
(C) Copyright 1988-1994 Paradigm Associates Inc.
10 heaps. size = 100000 cells, 2400000 bytes. 2048 inums. GC is mark and sweep
loading hello.scm
(define (main) (writes nil "Hello World
"))
done.
Hello World
EXIT
.EE

.SH RETURN VALUE
Errors during non-interactive usage mode result in a non-zero exit
status.

.SH ALTERNATIVE SYNTAX
Files which are loaded by the interpreter may contain the sequence
of characters parser:XXXXX in a comment in the first line of the file.
This will result in the automatic loading of the file parser_XXXXX.scm
which must provide a procedure parser_XXXXX to return a procedure
to read the forms from the file to be loaded. Useful values of XXXXX
include read, pratt, and fasl.

.SH FILES
/usr/local/lib/siod/*
/usr/local/lib/siod/siod.html
/usr/local/lib/libsiod.so
.PD
.SH SEE ALSO
command csiod

.I Structure and Interpretation of Computer Programs
, by Abelson and Sussman, MIT Press.

.SH DIAGNOSTICS
Error messages may also set the variable errobj to the offending object.
.SH BUGS
With -g1 it does not GC during EVAL, only before each READ/EVAL/PRINT cycle.
It does GC during EVAL with -g0, but that code may not run without modification
on all architectures.
.SH VERSION
Current version is 3.2, 12-MAR-96, by George Carrette. GJC\@world.std.com


tinyscheme1.39/CHANGES
     Change Log
     ----------
     Version 1.39
          Drew Yao fixed buffer overflow problems in mk_sharp_const.
     Version 1.38
          Interim release until the rewrite, mostly incorporating modifications from
          Kevin Cozens. Small addition for Cygwin in the makefile, and modifications
          by Andrew Guenther for Apple platforms.
     Version 1.37
          Joe Buehler submitted reserve_cells.
     Version 1.36
	  Joe Buehler fixed a patch in the allocator.
	  Alexander Shendi moved the comment handling in the scanner, which
	  fixed an obscure bug for which Mike E had provided a patch as well.
	  Kevin Cozens has submitted some fixes and modifications which have not
	  been incorporated yet in their entirety.
     Version 1.35
	  Todd Showalter discovered that the number of free cells reported
	  after GC was incorrect, which could also cause unnecessary allocations.
	 Version 1.34
	  Long missing version. Lots of bugfixes have accumulated in my email, so
	  I had to start using them. In this version, Keenan Pepper has submitted
	  a bugfix for the string comparison library procedure, Wouter Boeke
	  modified some code that was casting to the wrong type and crashed on
	  some machines, "SheppardCo" submitted a replacement "modulo" code and
	  Scott Fenton submitted lots of corrections that shut up some compiler
	  warnings. Brian Maher submitted instructions on how to build on OS-X.
	  I have to dig deeper into my mailbox and find earlier emails, too.
     Version 1.33
	  Charles Hayden fixed a nasty GC bug of the new stack frame, while in
	  the process of porting TinyScheme to C++. He also submitted other
	  changes, and other people also had comments or requests, but the GC
	  bug was so important that this version is put through the door to
	  correct it.
     Version 1.32
	  Stephen Gildea put some quality time on TinyScheme again, and made
	  a whole lot of changes to the interpreter that made it noticeably
	  faster.
     Version 1.31
          Patches to the hastily-done version 1.30. Stephen Gildea fixed
	  some things done wrongly, and Richard Russo fixed the makefile
	  for building on Windows. Property lists (heritage from MiniScheme)
	  are now optional and have dissappeared from the interface. They
	  should be considered as deprecated.
     Version 1.30
	  After many months, I followed Preston Bannister's advice of
	  using macros and a single source text to keep the enums and the
	  dispatch table in sync, and I used his contributed "opdefines.h".
	  Timothy Downs contributed a helpful function, "scheme_call".
	  Stephen Gildea contributed new versions of the makefile and
	  practically all other sources. He created a built-in STRING-APPEND,
	  and fixed a lot of other bugs.
	  Ruhi Bloodworth reported fixes necessary for OS X and a small
	  bug in dynload.c.
     Version 1.29
	  The previous version contained a lot of corrections, but there
	  were a lot more that still wait on a sheet of paper lost in a
	  carton someplace after my house move... Manuel Heras-Gilsanz
	  noticed this and resent his own contribution, which relies on
	  another bugfix that v.1.28 was missing: a problem with string
	  output, that this version fixes. I hope other people will take
	  the time to resend their contributions, if they didn't make it
	  to v.1.28.
     Version 1.28
	  Many people have contacted me with bugfixes or remarks in
	  the three months I was inactive. A lot of them spotted that
	  scheme_deinit crashed while reporting gc results. They suggested
	  that sc->outport be set to NIL in scheme_deinit, which I did.
	  Dennis Taylor remarked that OP_VALUEPRINT reset sc->value instead
	  of preserving it. He submitted a modification which I adopted
	  partially. David Hovemeyer sent me many little changes, that you
	  will find in version 1.28, and Partice Stoessel modified the
	  float reader to conform to R5RS.
     Version 1.27
          Version 1.27 is the successor of 1.25. Bug fixes only, but I had to
          release them so that everybody can profit. 'Backchar' tried to write
          back to the string, which obviously didn't work for const strings.
          'Substring' didn't check for crossed start and end indices. Defines
          changed to restore the ability to compile under MSVC.
     Version 1.26
          Version 1.26 was never released. I changed a lot of things, in fact
          too much, even the garbage collector, and hell broke loose. I'll
          try a more gradual approach next time.
     Version 1.25
          Types have been homogenized to be able to accomodate a different
	  representation. Plus, promises are no longer closures. Unfortunately,
	  I discovered that continuations and force/delay do not pass the SCM
	  test (and never did)... However, on the bright side, what little
	  modifications I did had a large impact on the footprint:
	  USE_NO_FEATURES now produces an object file of 63960 bytes on Linux!
     Version 1.24
	  SCM tests now pass again after change in atom2str.
     Version 1.23
          Finally I managed to mess it up with my version control. Version
	  1.22 actually lacked some of the things I have been fixing in the
	  meantime. This should be considered as a complete replacement for
	  1.22.
     Version 1.22
          The new ports had a bug in LOAD. MK_CLOSURE is introduced.
	  Shawn Wagner inquired about string->number and number->string.
	  I added string->atom and atom->string and defined the number
	  functions from them. Doing that, I fixed WRITE applied to symbols
	  (it didn't quote them). Unfortunately, minimum build is now
	  slightly larger than 64k... I postpone action because Jason's idea
	  might solve it elegantly.
     Version 1.21
          Jason Felice submitted a radically different datatype representation
	  which he had implemented. While discussing its pros and cons, it
	  became apparent that the current implementation of ports suffered

siod/siod.c
/* Hyper Text Query Server
   23-DEC-94 George J. Carrette.

 This is a SIOD main program with additional command-line processing
 functionality.

*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "siod.h"

static char *siod_argv[] = {
  "siod",
  "-h100000:10",
  "-g0",
  "-o1000",
  "-s200000",
  "-n2048"};

static void init_htqs_version(void)
{setvar(cintern("*htqs-version*"),
	cintern("$Id: siod.c,v 1.6 1996/03/10 02:31:10 gjc Exp $"),
	NIL);}

static LISP cgi_main(LISP result)
{if (CONSP(result) && TYPEP(car(result),tc_string))
   {put_st("Status: 500 Server Error (Application)\n");
    put_st("Content-type: text/html\n\n");
    put_st("<HTML><HEAD><TITLE>Server Error (Application)</TITLE></HEAD>\n");
    put_st("<BODY><H1>Server Error (Application)</H1>\n");
    put_st("An application on this server has encountered an error\n");
    put_st("which prevents it from fulfilling your request.");
    put_st("<P><PRE><B>Error Message:</B> ");
    lprint(car(result),NIL);
    if NNULLP(cdr(result))
      {put_st("\n");
       lprint(cdr(result),NIL);}
    put_st("</PRE></BODY></HTML>\n");
    err("cgi-main",NIL);}
 return(NIL);}

main(int argc,char **argv
#if defined(unix) || defined(vms)
     ,char **env
#endif
     )
{int j,retval,iargc,mainflag = 0,text_plain_flag = 0;
 char *iargv[2],*start,*end;
 LISP l;
 process_cla(sizeof(siod_argv)/sizeof(char *),siod_argv,1);
 iargv[0] = "";
 for(iargc=0,j=1;j<argc; ++j)
   if (*(start = argv[j]) == '-')
     {while(*start)
	{if (!(end = strstr(start,",-"))) end = &start[strlen(start)];
	 iargv[1] = (char *) malloc(end-start+1);
	 memcpy(iargv[1],start,end-start);
	 iargv[1][end-start] = 0;
	 if ((strncmp(iargv[1],"-v",2) == 0) &&
	     (atol(&iargv[1][2]) > 0) &&
	     (iargv[1][2] != '0'))
	   {printf("Content-type: text/plain\r\n\r\n");
	    text_plain_flag = 1;}
	 if ((strncmp(iargv[1],"-m",2) == 0))
	   mainflag = atol(&iargv[1][2]);
	 else
	   process_cla(2,iargv,1);
	 /* Note: Not doing free(iargv[1]); */
	 start = (*end) ? end+1 : end;}}
 else
   ++iargc;
 print_welcome();
 print_hs_1();
 init_storage();
 for(l=NIL,j=0;j<argc;++j)
   l = cons(strcons(strlen(argv[j]),argv[j]),l);
 setvar(cintern("*args*"),nreverse(l),NIL);
 l = NIL;
#if defined(unix) || defined(vms)
 for(l=NIL,j=0;env[j];++j)
   l = cons(strcons(strlen(env[j]),env[j]),l);
 setvar(cintern("*env*"),nreverse(l),NIL);
 l = NIL;
#endif
 init_subrs();
 init_trace();
 init_slibu();
 init_subr_1("__cgi-main",cgi_main);
 init_htqs_version();

 if (iargc == 0)
   retval = repl_driver(1,1,NULL);
 else
   {for(j=1;j<(((mainflag >= 2) && (argc > 3)) ? 3 : argc);++j)
      if (argv[j][0] != '-')
	{retval = htqs_arg(argv[j]);
	 if (retval != 0) break;}
    if (mainflag)
      retval = htqs_arg(((mainflag > 2) && !text_plain_flag)
			? "(__cgi-main (*catch 'errobj (main))))"
			: "(main)");}
 if (siod_verbose_check(2))
   printf("EXIT\n");
 exit(retval);}

int htqs_arg(char *value)
{char tmpbuff[256];
 if ((strcmp(value,"(repl)") == 0) ||
     (strcmp(value,"repl") == 0))
   return(repl_driver(1,1,NULL));
 else if (!strchr(value,'('))
   {sprintf(tmpbuff,"(require \"%s\")",value);
    return(repl_c_string(tmpbuff,0,0,0));}
 else
   return(repl_c_string(value,0,0,0));}

